src/analyzer/protocol/modbus/modbus-analyzer.pac

```diff
@@ -355,6 +355,154 @@ refine flow ModbusTCP_Flow += {
 		%}
 
 
+	# REQUEST FC=8
+	function deliver_DiagnosticsRequest(header: ModbusTCP_TransportHeader, message: DiagnosticsRequest): bool
+		%{
```

Comment:
As above, if we used a lower-level event, the conversions (and possibly also weird generations) could be done in Zeek script. Think the analyzer shouldn't have as much logic as it has now in the request/response functions and rather do this in the script layer.

The weird generation could possibly be converted to `&enforce` annotations and actually trigger analyzer violations or not encoded?

Mainly that we're not raising events if something is weird based on rules hard-coded in the analyzer seems a bit unfortunate.

Reply:
> The weird generation could possibly be converted to &enforce annotations and actually trigger analyzer violations or not encoded?

This is tough to do because the various validations depend what subfunction is selected. I could split it up into individual types in the .pac file, but that seems overkill when I can do the same validation on the C++ side.

> Mainly that we're not raising events if something is weird based on rules hard-coded in the analyzer seems a bit unfortunate.

The weirds being raised are based on the spec itself, so I'm not sure what you mean.

Reply:
> The weirds being raised are based on the spec itself, so I'm not sure what you mean.

Would just always raise the `modbus_diagnostics_request` and response event, even if not spec conform but otherwise parsable.

A non-conformant  or development device with a new subfunction code added would not generate a diagnostics, running minimally into the `default` case even if structurally the request/response is valid and could be raised to script-land.

```
428                                 default:                                                                                                                                                                                                                                                                      
429                                         zeek::reporter->Weird("modbus_diag_unknown_request_subfunction",                                                                                                                                                                                                      
430                                                               zeek::util::fmt("%d", ${message.subfunction}));                                                                                                                                                                                                 
431                                         return false;                                                                                                                                                                                                                                                         
432                                 } 
```

If data format of known subfunction codes doesn't agree with spec, still possibly interesting to get those event into script-land. Also easier to extend/change the policy the in script layer - thinking of a `modbus_diagnostics.log` or so.

Reply:
I don't feel super strongly about it, but we do have a bunch of prior places where we do the same thing with other analyzers (see, for example, all of the weirds reported by the GRE analyzer). They're not checking for violations of the parsing, but for violations of the data encountered while parsing, and bailing out of the analyzer at that point.

Reply:
Me neither. GRE might be historical due to being a packet analyzer? Compromise could be to generate the weird but fall through to raising the event regardless.

Then, if there's feedback that either isn't wanted (or weird generation in script-land allows easier customization) revisit.

Leaving it to you :-)